### PR TITLE
build: refine docker image CI process

### PR DESCRIPTION
**Pull request by mattklein123**
Fixes https://github.com/envoyproxy/envoy/issues/10716

### CODE REVIEW

**Comment by mattklein123** [comment]
@lizan I tested various permutations of this locally with different env var combinations so I'm fairly confident it works correctly, but would definitely appreciate a close look.

File: ci/docker_ci.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  IMAGE_NAME="${AZP_BRANCH/refs\/tags\//}"
+fi
+
+docker login -u "$DOCKERHUB_USERNAME" -p "$DOCKERHUB_PASSWORD"
+
+for BUILD_TYPE in "" "-alpine" "-alpine-debug"; do
+    docker tag "${DOCKER_IMAGE_PREFIX}${BUILD_TYPE}:local" "${DOCKER_IMAGE_PREFIX}${BUILD_TYPE}${IMAGE_POSTFIX}:${IMAGE_NAME}"
+    docker push "${DOCKER_IMAGE_PREFIX}${BUILD_TYPE}${IMAGE_POSTFIX}:${IMAGE_NAME}"
+
+    docker tag "${DOCKER_IMAGE_PREFIX}${BUILD_TYPE}:local" "${DOCKER_IMAGE_PREFIX}${BUILD_TYPE}${IMAGE_POSTFIX}:latest"
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply ci/docker_ci.sh:35]: I thought about this. I agree it makes sense to not do any latest for non dev. I will make that change.

File: ci/docker_ci.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+DOCKER_IMAGE_PREFIX="${DOCKER_IMAGE_PREFIX:-envoyproxy/envoy}"
+
+# Test the docker build in all cases, but use a local tag that we will overwrite before push in the
+# cases where we do push.
+for BUILD_TYPE in "" "-alpine" "-alpine-debug"; do
+    docker build -f ci/Dockerfile-envoy"${BUILD_TYPE}" -t "${DOCKER_IMAGE_PREFIX}${BUILD_TYPE}:local" .
+done
+
+# Only push images for master builds and tag builds.
+if [[ "${AZP_BRANCH}" != 'refs/heads/master' ]] && ! [[ "${AZP_BRANCH}" =~ ^refs/tags/v.* ]]; then
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply ci/docker_ci.sh:16]: I don't think we actually want that. Any stable release would be tagged, right? And that is covered by this tag clause. Tag CI runs the release build, so it should all work out independently.

**Comment by mattklein123** [comment]
> Also can you apply following so we run postsubmit CI against release branch?

Yes I will make the change, but see my other comment. I don't think we want to push images for stable branch builds, only for tags. This would be useful for CI though.

**Comment by mattklein123** [comment]
@lizan updated PTAL. I did not make the one change your requested to push images on release branch build since IMO we don't want that.

File: ci/docker_ci.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (lizan) [reply ci/docker_ci.sh:16]: We do want that. If we don't run postsubmit on release branch, we can't know whether there's a merge race in stable branch until we tag, as well as in the security repo, which is too late.

[the remark this answers is not in the document]
Comment (mattklein123) [reply ci/docker_ci.sh:16]: We are running postsubmit now on the branch, we just aren't pushing images... Please check the new diff.

**Comment by mattklein123** [comment]
@lizan updated per offline convo

**Pull request merged**